File: examples/platforms/android/qml_in_java_based_android_project/app/src/main/java/com/example/qml_in_java_based_android_project/MainActivity.java
// Copyright (C) 2024 The Qt Company Ltd.
// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR BSD-3-Clause
package com.example.qml_in_java_based_android_project;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SwitchCompat;
import android.content.res.Configuration;
import android.graphics.Color;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.qtproject.qt.android.QtQuickView;

import java.util.HashMap;
import java.util.Map;


// Implement QtQuickView StatusChangeListener interface to get status updates
// from the underlying QQuickView
public class MainActivity extends AppCompatActivity implements QtQuickView.StatusChangeListener {

    private static final String TAG = "myTag";
    private final Colors m_colors = new Colors();
    private final Map<Integer, String> m_statusNames = new HashMap<Integer, String>()  {{
        put(QtQuickView.STATUS_READY, " READY");
        put(QtQuickView.STATUS_LOADING, " LOADING");
        put(QtQuickView.STATUS_ERROR, " ERROR");
        put(QtQuickView.STATUS_NULL, " NULL");
    }};
    private int m_qmlButtonSignalListenerId;
    private LinearLayout m_mainLinear;
    private FrameLayout m_qmlFrameLayout;
    private QtQuickView m_qmlView;
    private LinearLayout m_androidControlsLayout;
    private TextView m_getPropertyValueText;
    private TextView m_qmlStatus;
    private SwitchCompat m_switch;
    private View m_box;

    //! [onCreate]
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        m_mainLinear = findViewById(R.id.mainLinear);
        m_getPropertyValueText = findViewById(R.id.getPropertyValueText);
        m_qmlStatus = findViewById(R.id.qmlStatus);
        m_androidControlsLayout = findViewById(R.id.javaLinear);
        m_box = findViewById(R.id.box);

        m_switch = findViewById(R.id.switch1);
        m_switch.setOnClickListener(view -> switchListener());
        //! [m_qmlView]
        m_qmlView = new QtQuickView(this, "qrc:/qt/qml/qml_in_android_view/main.qml",
                "qml_in_android_view");
        //! [m_qmlView]

        // Set status change listener for m_qmlView
        // listener implemented below in OnStatusChanged
        //! [setStatusChangeListener]
        m_qmlView.setStatusChangeListener(this);
        //! [setStatusChangeListener]
        //! [layoutParams]
        ViewGroup.LayoutParams params = new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        m_qmlFrameLayout = findViewById(R.id.qmlFrame);
        m_qmlFrameLayout.addView(m_qmlView, params);
        //! [layoutParams]
        Button button = findViewById(R.id.button);
        button.setOnClickListener(view -> onClickListener());

        // Check target device orientation on launch
        handleOrientationChanges();
    }
    //! [onCreate]
    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        handleOrientationChanges();
    }

    private void handleOrientationChanges() {
        // When specific target device display configurations (listed in AndroidManifest.xml
        // android:configChanges) change, get display metrics and make needed changes to UI
        DisplayMetrics displayMetrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        ViewGroup.LayoutParams qmlFrameLayoutParams = m_qmlFrameLayout.getLayoutParams();
        ViewGroup.LayoutParams linearLayoutParams = m_androidControlsLayout.getLayoutParams();

        if (displayMetrics.heightPixels > displayMetrics.widthPixels) {
            m_mainLinear.setOrientation(LinearLayout.VERTICAL);
            qmlFrameLayoutParams.width = ViewGroup.LayoutParams.MATCH_PARENT;
            qmlFrameLayoutParams.height = 0;
            linearLayoutParams.width = ViewGroup.LayoutParams.MATCH_PARENT;
            linearLayoutParams.height = 0;
        } else {
            m_mainLinear.setOrientation(LinearLayout.HORIZONTAL);
            qmlFrameLayoutParams.width = 0;
            qmlFrameLayoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT;
            linearLayoutParams.width = 0;
            linearLayoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT;
        }
        m_qmlFrameLayout.setLayoutParams(qmlFrameLayoutParams);
        m_androidControlsLayout.setLayoutParams(linearLayoutParams);
    }

    //! [onStatusChanged]
    @Override
    public void onStatusChanged(int status) {
        Log.i(TAG, "Status of QtQuickView: " + status);

        final String qmlStatus = getResources().getString(R.string.qml_view_status)
                + m_statusNames.get(status);

        // Show current QML View status in a textview
        m_qmlStatus.setText(qmlStatus);

        // Connect signal listener to "onClicked" signal from main.qml
        // addSignalListener returns int which can be used later to identify the listener
        //! [qml signal listener]
        if (status == QtQuickView.STATUS_READY && !m_switch.isChecked()) {
            m_qmlButtonSignalListenerId = m_qmlView.connectSignalListener("onClicked", Object.class,
                    (String signal, Object o) -> {
                Log.i(TAG, "QML button clicked");
                m_androidControlsLayout.setBackgroundColor(Color.parseColor(m_colors.getColor()));
            });

        }
        //! [qml signal listener]
    }
    //! [onStatusChanged]
    //! [onClickListener]
    public void onClickListener() {
        // Set the QML view root object property "colorStringFormat" value to
        // color from Colors.getColor()
        m_qmlView.setProperty("colorStringFormat", m_colors.getColor());

        String qmlBackgroundColor = m_qmlView.getProperty("colorStringFormat");

        // Display the QML View background color code
        m_getPropertyValueText.setText(qmlBackgroundColor);

        // Display the QML View background color in a view
        m_box.setBackgroundColor(Color.parseColor(qmlBackgroundColor));
    }
    //! [onClickListener]

    public void switchListener() {
        TextView text = findViewById(R.id.switchText);
        // Disconnect QML button signal listener if switch is On using the saved signal listener Id
        // and connect it again if switch is turned off
        if (m_switch.isChecked()) {
            Log.i(TAG, "QML button onClicked signal listener disconnected");
            text.setText(R.string.connect_qml_button_signal_listener);
            //! [disconnect qml signal listener]
            m_qmlView.disconnectSignalListener(m_qmlButtonSignalListenerId);
            //! [disconnect qml signal listener]
        } else {
            Log.i(TAG, "QML button onClicked signal listener connected");
            text.setText(R.string.disconnect_qml_button_signal_listener);
            m_qmlButtonSignalListenerId = m_qmlView.connectSignalListener("onClicked",
                    Object.class, (String t, Object value) -> {
                Log.i(TAG, "QML button clicked");
                m_androidControlsLayout.setBackgroundColor(Color.parseColor(m_colors.getColor()));
            });
        }
    }
}

File: examples/platforms/android/qml_in_kotlin_based_android_project/app/src/main/java/com/example/qml_in_kotlin_based_android_project/MainActivity.kt
package com.example.qml_in_kotlin_based_android_project

import android.content.res.Configuration
import android.graphics.Color
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.Log
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import com.example.qml_in_kotlin_based_android_project.databinding.ActivityMainBinding
import org.qtproject.example.qml_in_android_view.QmlModule.Main
import org.qtproject.qt.android.QtQmlStatus
import org.qtproject.qt.android.QtQmlStatusChangeListener
import org.qtproject.qt.android.QtQuickView

class MainActivity : AppCompatActivity(), QtQmlStatusChangeListener {

    private val TAG = "myTag"
    private val m_colors: Colors = Colors()
    private lateinit var m_binding: ActivityMainBinding
    private var m_qmlButtonSignalListenerId = 0
    private var m_qtQuickView: QtQuickView? = null
    //! [qmlComponent]
    private var m_mainQmlComponent: Main = Main()
    //! [qmlComponent]
    private val m_statusNames = hashMapOf(
        QtQmlStatus.READY to "READY",
        QtQmlStatus.LOADING to "LOADING",
        QtQmlStatus.ERROR to "ERROR",
        QtQmlStatus.NULL to "NULL"
    )
    //! [onCreate]
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //! [binding]
        m_binding = ActivityMainBinding.inflate(layoutInflater)
        val view = m_binding.root
        setContentView(view)
        //! [binding]

        m_binding.signalSwitch.setOnClickListener { switchListener() }

        //! [m_qtQuickView]
        m_qtQuickView = QtQuickView(this)
        //! [m_qtQuickView]

        // Set status change listener for m_qmlView
        // listener implemented below in OnStatusChanged
        //! [setStatusChangeListener]
        m_mainQmlComponent.setStatusChangeListener(this)
        //! [setStatusChangeListener]

        //! [layoutParams]
        val params: ViewGroup.LayoutParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
        )
        m_binding.qmlFrame.addView(m_qtQuickView, params)
        //! [layoutParams]
        //! [loadComponent]
        m_qtQuickView!!.loadComponent(m_mainQmlComponent)
        //! [loadComponent]

        m_binding.changeColorButton.setOnClickListener { onClickListener() }

        // Check target device orientation on launch
        handleOrientationChanges()
    }
    //! [onCreate]
    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        handleOrientationChanges()
    }

    private fun handleOrientationChanges() {
        // When specific target device display configurations (listed in AndroidManifest.xml
        // android:configChanges) change, get display metrics and make needed changes to UI
        val displayMetrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(displayMetrics)
        val qmlFrameLayoutParams = m_binding.qmlFrame.layoutParams
        val linearLayoutParams = m_binding.kotlinLinear.layoutParams

        if (displayMetrics.heightPixels > displayMetrics.widthPixels) {
            m_binding.mainLinear.orientation = LinearLayout.VERTICAL
            qmlFrameLayoutParams.width = ViewGroup.LayoutParams.MATCH_PARENT
            qmlFrameLayoutParams.height = 0
            linearLayoutParams.width = ViewGroup.LayoutParams.MATCH_PARENT
            linearLayoutParams.height = 0
        } else {
            m_binding.mainLinear.orientation = LinearLayout.HORIZONTAL
            qmlFrameLayoutParams.width = 0
            qmlFrameLayoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT
            linearLayoutParams.width = 0
            linearLayoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT
        }
        m_binding.qmlFrame.layoutParams = qmlFrameLayoutParams
        m_binding.kotlinLinear.layoutParams = linearLayoutParams
    }

    //! [onClickListener]
    private fun onClickListener() {
        // Set the QML view root object property "colorStringFormat" value to
        // color from Colors.getColor()
        m_mainQmlComponent.colorStringFormat = m_colors.getColor()

        val qmlBackgroundColor = m_mainQmlComponent.colorStringFormat

        // Display the QML View background color code
        m_binding.getPropertyValueText.text = qmlBackgroundColor

        // Display the QML View background color in a view
        m_binding.colorBox.setBackgroundColor(Color.parseColor(qmlBackgroundColor))
    }
    //! [onClickListener]

    private fun switchListener() {
        // Disconnect QML button signal listener if switch is On using the saved signal listener Id
        // and connect it again if switch is turned off
        if (m_binding.signalSwitch.isChecked) {
            Log.v(TAG, "QML button onClicked signal listener disconnected")
            m_binding.switchText.setText(R.string.connect_qml_button_signal_listener)
            //! [disconnect qml signal listener]
            m_mainQmlComponent.disconnectSignalListener(m_qmlButtonSignalListenerId)
            //! [disconnect qml signal listener]
        } else {
            Log.v(TAG, "QML button onClicked signal listener connected")
            m_binding.switchText.setText(R.string.disconnect_qml_button_signal_listener)
            m_qmlButtonSignalListenerId =
                m_mainQmlComponent.connectOnClickedListener { _: String, _: Void? ->
                    Log.i(TAG, "QML button clicked")
                    m_binding.kotlinLinear.setBackgroundColor(
                        Color.parseColor(
                            m_colors.getColor()
                        )
                    )
                }
        }
    }

    //! [onStatusChanged]
    override fun onStatusChanged(status: QtQmlStatus?) {
        Log.v(TAG, "Status of QtQuickView: $status")

        val qmlStatus = (resources.getString(R.string.qml_view_status)
                + m_statusNames[status])

        // Show current QML View status in a textview
        m_binding.qmlStatus.text = qmlStatus

        // Connect signal listener to "onClicked" signal from main.qml
        // addSignalListener returns int which can be used later to identify the listener
        //! [qml signal listener]
        if (status == QtQmlStatus.READY && !m_binding.signalSwitch.isChecked) {
            m_qmlButtonSignalListenerId =
                m_mainQmlComponent.connectOnClickedListener { _: String, _: Void? ->
                    Log.i(TAG, "QML button clicked")
                    m_binding.kotlinLinear.setBackgroundColor(
                        Color.parseColor(
                            m_colors.getColor()
                        )
                    )
                }
        }
        //! [qml signal listener]
    }
    //! [onStatusChanged]
}
